### PR TITLE
cli: Add tag-aware upgrade operations

### PR DESCRIPTION
**Pull request by gursewak1997**
Implements `bootc upgrade --tag <TAG>` to upgrade to a different tag
of the current image without typing the full registry path.

Implementation:
- Added `ImageReference::with_tag()` method in spec.rs
- Uses OCI Reference API for registry and containers-storage transports
- Falls back to string manipulation for oci: and other complex transports
- Properly strips digests when replacing tags
- Works with all transport types

Testing:
- Unit tests cover registry, containers-storage, digest handling, and edge cases
- Integration test (test-upgrade-tag.nu) uses containers-storage to verify
  full tag-switching workflow across reboots without registry dependency

### CODE REVIEW

File: crates/lib/src/cli.rs [hunk 1]
```diff
@@ -123,6 +125,17 @@ pub(crate) struct UpgradeOpts {
...
+    /// Upgrade to a different tag of the currently booted image.
+    ///
+    /// This derives the target image by replacing the tag portion of the current
+    /// booted image reference. Only works when booted from a registry image.
+    #[clap(long)]
+    pub(crate) tag: Option<String>,
+
+    /// List available tags for the currently booted image repository.
+    #[clap(long, conflicts_with_all = ["apply", "download_only", "from_downloaded"])]
+    pub(crate) list_tags: bool,
```
Comment (cgwalters) [inline crates/lib/src/cli.rs]: This one feels like it should be a separate operation.

File: crates/lib/src/cli.rs [hunk 2]
```diff
@@ -1220,6 +1253,70 @@ async fn upgrade(
     Ok(())
 }
 
+/// Derive a new image reference by replacing the tag of the current image
+fn derive_image_with_tag(current: &ImageReference, new_tag: &str) -> Result<ImageReference> {
+    // Only works for registry transport
```
Comment (cgwalters) [inline crates/lib/src/cli.rs]: I think we can easily support non-registry transports here.

File: crates/lib/src/cli.rs [hunk 3]
```diff
@@ -1220,6 +1253,70 @@ async fn upgrade(
...
+        );
+    }
+
+    // Parse the current image reference
+    let reference: oci_spec::distribution::Reference = current.image.parse()?;
+
+    // Build new image reference with the new tag (stripping any digest)
+    let registry = reference.registry();
+    let repository = reference.repository();
+    let new_image = format!("{}/{}:{}", registry, repository, new_tag);
```
Comment (cgwalters) [inline crates/lib/src/cli.rs]: There's an API for this https://docs.rs/oci-spec/latest/oci_spec/distribution/struct.Reference.html#method.with_tag

File: crates/lib/src/cli.rs [hunk 4]
```diff
@@ -1220,6 +1253,70 @@ async fn upgrade(
...
+
+    // Construct repository name: registry + repository
+    let registry = reference.registry();
+    let repository = reference.repository();
+    let repo_name = format!("{}/{}", registry, repository);
+
+    let repo = format!("docker://{}", repo_name);
+
+    // Use skopeo to list tags
+    let output = tokio::process::Command::new("skopeo")
```
Comment (cgwalters) [inline crates/lib/src/cli.rs]: One problem is this won't use the ostree/bootc pull secret by default; we have a whole containers-image-proxy abstraction that we need to use in general for this.

Though the next problem here is that code doesn't expose an abstraction for this...but should.

File: crates/lib/src/cli.rs [hunk 5]
```diff
@@ -1220,6 +1253,70 @@ async fn upgrade(
...
+        .output()
+        .await
+        .context("Failed to execute skopeo")?;
+
+    if !output.status.success() {
+        let stderr = String::from_utf8_lossy(&output.stderr);
+        anyhow::bail!("Failed to list tags: {}", stderr);
+    }
+
+    let stdout = String::from_utf8_lossy(&output.stdout);
```
Comment (cgwalters) [inline crates/lib/src/cli.rs]: For things like this it would be a lot cleaner to use `skopeo inspect` which outputs JSON. I think it's almost always wrong to use `from_utf8_lossy`.

Perhaps this might be best done with `upgrade --check`? We could also cache the latest available tags there.

File: crates/lib/src/cli.rs [hunk 6]
```diff
@@ -2203,6 +2300,78 @@ mod tests {
...
+            Opt::Upgrade(opts) => {
+                assert_eq!(opts.tag, Some("v1.1".to_string()));
+                assert!(opts.check);
+            }
+            _ => panic!("Expected Upgrade variant"),
+        }
+    }
+
+    #[test]
+    fn test_derive_image_with_tag() {
```
Comment (cgwalters) [inline crates/lib/src/cli.rs]: Nothing wrong with unit tests (although this is basically reimplementing the `with_tag` API I mentioned) - but we really want an integration (tmt) test for this. Right now our default integration tests don't involve a registry (that's a different issue) but I think we can do the same with `containers-storage:` - basically bootc switch to one tag and then switch to a different one etc.

File: crates/lib/src/cli.rs (hunk 5 above)
Comment (gursewak1997) [reply crates/lib/src/cli.rs]: I've switched to skopeo inspect with JSON parsing (parsing the RepoTags field) instead of from_utf8_lossy. Also integrated it with `upgrade --check` as an optional `--show-tags` flag.
Caching the tags during check is a good idea. I can add that if you think it's worth the added complexity.

File: crates/lib/src/cli.rs (hunk 6 above)
Comment (gursewak1997) [reply crates/lib/src/cli.rs]: Added

File: crates/lib/src/cli.rs [hunk 7]
```diff
@@ -1220,6 +1263,87 @@ async fn upgrade(
     Ok(())
 }
 
+/// Derive a new image reference by replacing the tag of the current image
+fn derive_image_with_tag(current: &ImageReference, new_tag: &str) -> Result<ImageReference> {
```
Comment (cgwalters) [inline crates/lib/src/cli.rs]: Hmm I think this should probably be a method on `ImageReference` instead of being all the way at the top in the CLI code.

That said I am pretty sure for `containers-storage` we can also create a `Reference` instead of reimplementing parsing.

The complicated one is `oci` because it's filesystem paths and `:` creates ambiguity; there's some complex code IIRC in containers/container-libs around this in Go which we may need to replicate. I think this came up elsewhere.

File: crates/lib/src/cli.rs (hunk 4 above)
Comment (cgwalters) [reply crates/lib/src/cli.rs]: You marked this resolved but I don't think it is right? IOW if the image is behind a pull secret then this will fail. Look at the containers-image-proxy code that we use. We need to pass the ostree auth file to skopeo here.

Comment (cgwalters) [reply crates/lib/src/cli.rs]: Bigger picture can we just drop this list-tags functionality for now and instead suggest to users to directly run `skopeo inspect` on their own?

I know it's less ergonomic and this whole thing is just about ergonomics, but hopefully the `--tag` is enough.

My worry is basically again:

- We must handle authfiles and such using the ostree/bootc one
- This feels like a kind of scope creep into general "inspect the remote container" reference but users can already just run `skopeo` with their own options if that's what they want.

Maybe we could add `skopeo inspect --booted` for example? But that still doesn't solve the authfile discrepancy.

(I know the authfile thing is also quite annoying...see also https://github.com/containers/image/pull/2600 )

Comment (gursewak1997) [reply crates/lib/src/cli.rs]: I see. Dropped the `list-tags` functionality for now
